src/session.c

```diff
@@ -421,18 +463,39 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
         }
         if (pin) {
             pinlen = strlen((const char *)pin);
-        } else if (pw_cb) {
-            const char *login_info = p11prov_slot_get_login_info(slot);
-            OSSL_PARAM params[2] = {
-                OSSL_PARAM_DEFN(OSSL_PASSPHRASE_PARAM_INFO,
-                                OSSL_PARAM_UTF8_STRING, (void *)login_info,
-                                strlen(login_info)),
-                OSSL_PARAM_END,
-            };
-            ret = pw_cb(cb_pin, sizeof(cb_pin), &cb_pin_len, params, pw_cbarg);
-            if (ret != RET_OSSL_OK) {
-                ret = CKR_GENERAL_ERROR;
-                goto done;
+        } else {
+            if (pw_cb) {
+                const char *login_info = p11prov_slot_get_login_info(slot);
+                OSSL_PARAM params[2] = {
+                    OSSL_PARAM_DEFN(OSSL_PASSPHRASE_PARAM_INFO,
+                                    OSSL_PARAM_UTF8_STRING, (void *)login_info,
+                                    strlen(login_info)),
+                    OSSL_PARAM_END,
+                };
+                ret = pw_cb(cb_pin, sizeof(cb_pin), &cb_pin_len, params,
+                            pw_cbarg);
+                if (ret != RET_OSSL_OK) {
+                    /* this error can mean anything from the user canceling
+                     * the prompt to no UI method provided.
+                     * Fall back to our prompt here */
+                    ret = p11prov_session_get_pin(slot, (char *)cb_pin,
+                                                  &cb_pin_len);
+                    if (ret != RET_OSSL_OK) {
+                        /* give up */
+                        ret = CKR_GENERAL_ERROR;
+                        goto done;
+                    }
+                }
+            } else {
+                /* We are asking the user off-band for the user consent -- from
+                 * store we will always receive non-null (but unusable) callback
```

Comment:
Not sure why this is unusable? When we receive a callback it is always usable, the problem is that often we just do not have one, and we can't store the callback we were provided for the store operation for later operations as the calling application may not expect it.

Reply:
As discussed above, from the store we will always get a callback (at least in OpenSSL in Fedora 38), but the callback just fails with the error mentioned above.

Reply:
Its the conversion from the UI methods on the store API to the low-level callbacks on the provider level. What you describe would make sense and I would be assuming this, but this is OpenSSL. So when the store API is called without the UI method, the provider gets non-null callback, but that one fails because there is no assigned UI method. Sounds as stupid as it is (or I followed the gdb wrongly -- there is obviously no documentation. You can double-check that this wont work if you remove this code branch).